src/NextGen/DonationForm/Controllers/DonateController.php

```diff
@@ -92,37 +92,52 @@ private function getOrCreateDonor(
     }
 
     /**
-     * This logic is intended to work with the legacy receipt functionality
-     * by setting and updating the give_purchase session.
-     *
      * @unreleased
      *
      * @return void
      */
-    private function setSession(Donation $donation, Donor $donor)
+    private function saveCustomFields(DonationForm $form, Donation $donation, array $customFields)
     {
-        give()->session->maybe_start_session();
-
-        $purchaseSession = (array)give()->session->get('give_purchase');
-
-        if ($purchaseSession && array_key_exists('donation_id', $purchaseSession)) {
-            $purchaseSession['donation_id'] = $donation->id;
-
-            give()->session->set('give_purchase', $purchaseSession);
-        } else {
-            $legacyPurchaseFormData = LegacyPurchaseFormData::fromArray(['donation' => $donation, 'donor' => $donor]);
+        (new StoreCustomFields())($form, $donation, $customFields);
+    }
 
-            give_set_purchase_session($legacyPurchaseFormData->toPurchaseData());
-        }
+    /**
+     * Use our new receipt url for the success page uri
+     *
+     * @unreleased
+     *
+     * @return void
+     */
+    protected function filterSuccessPageUri(string $filteredUrl)
+    {
+        add_filter('give_get_success_page_uri', static function ($url) use ($filteredUrl) {
```

Comment:
In general, should we be adding filters like this inside of the controller? Usually filters are added to a Service Provider. This would be a big break in convention if we start including filters within controllers.

Reply:
I see why you put the filter in the controller, but I think it warrants a discussion.

Reply:
Yes I realize this is not ideal, It's tricky because we need to completely replace that value based on the current request - that would not be available in a service provider.  I should note that this is not intended to be for the long term, eventually we will want to update the gateway api to handle redirecting to the appropriate success page.

Reply:
I agree it's not ideal, but so long as the API for the gateways themselves is declarative and immutable, stuff like this doesn't bother me too much. Make sure to leave a note here that as to the fact that it's temporary and how it ideally ought to be handled in the future.